[Bug]: DynamoDB table tagging failures due to API changes

### Terraform Core Version

1.9.5

### AWS Provider Version

5.66.0

### Affected Resource(s)

* aws_dynamodb_tag
* aws_dynamodb_table

### Expected Behavior

Adding or removing tags w/ aws_dynamodb_tag and aws_dynamodb_table should work without error.

### Actual Behavior

Various errors including `Error: reading dynamodb resource (arn:aws:dynamodb:us-west-2:XXX:table/test) tag (t1): empty result` and `SubscriberLimitExceeded`

### Relevant Error/Panic Output Snippet

```shell
aws_dynamodb_tag.tag3: Creating...
aws_dynamodb_tag.tag1: Creating...
aws_dynamodb_tag.tag2: Creating...
aws_dynamodb_tag.tag2: Still creating... [10s elapsed]
╷
│ Error: reading dynamodb resource (arn:aws:dynamodb:us-west-2:XXXXX:table/test) tag (t1): empty result
│
│   with aws_dynamodb_tag.tag1,
│   on blueprint.tf line 38, in resource "aws_dynamodb_tag" "tag1":
│   38: resource "aws_dynamodb_tag" "tag1" {
│
╵
╷
│ Error: reading dynamodb resource (arn:aws:dynamodb:us-west-2:XXXX:table/test) tag (t2): empty result
│
│   with aws_dynamodb_tag.tag2,
│   on blueprint.tf line 44, in resource "aws_dynamodb_tag" "tag2":
│   44: resource "aws_dynamodb_tag" "tag2" {
│
╵
╷
│ Error: reading dynamodb resource (arn:aws:dynamodb:us-west-2:XXXX:table/test) tag (t3): empty result
│
│   with aws_dynamodb_tag.tag3,
│   on blueprint.tf line 50, in resource "aws_dynamodb_tag" "tag3":
│   50: resource "aws_dynamodb_tag" "tag3" {
```


### Terraform Configuration Files

terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.66.0"
    }
  }
}

provider "aws" {
  region = "us-west-2"
  max_retries = "10"
}

resource "aws_dynamodb_table" "test" {
  name           = "test"
  billing_mode = "PAY_PER_REQUEST"
  hash_key       = "id"

  attribute {
    name = "id"
    type = "S"
  }
}

resource "aws_dynamodb_tag" "tag1" {
  resource_arn = aws_dynamodb_table.test.arn
  key          = "t1"
  value        = "v1"
}

resource "aws_dynamodb_tag" "tag2" {
  resource_arn = aws_dynamodb_table.test.arn
  key          = "t2"
  value        = "v2"
}

resource "aws_dynamodb_tag" "tag3" {
  resource_arn = aws_dynamodb_table.test.arn
  key          = "t3"
  value        = "v3"
}


### Steps to Reproduce

Apply the above terraform.

Around August 23, 2024, AWS changed TagResource and UntagResource APIs to be asynchronous / eventually consistent. Previously these API calls were blocking / strongly consistent. This causes at least two failure modes.

1) Applying the terraform included above exposes the first problem, failing with `Error: reading dynamodb resource`. After successfully calling TagResource, Terraform fails to read back the tag value. Presumably this is because the eventually consistent write has not yet applied, so the read comes back empty.

2) A second problem also occurs but is masked by retries. AWS does not allow TagResource to be called until the previous asynchronous call has completed. When applying the three tags in the blueprint above, it results in three calls to TagResource. Some of these calls will fail with a `SubscriberLimitExceeded` error message but are eventually retried. You can see this by enabling DEBUG output and looking for `SubscriberLimitExceeded`. With enough retries these eventually succeed, as the previous in-flight calls complete. You can also reproduce this error by setting `max_retries = 1` or by applying with version 5.35.0 or older, which does not retry and fails immediately.


### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

None

Acceptance tests for this resource are also now failing. Generated tagging logic will need to be updated to account for eventual consistency of the `ListTags` result.

```console
% make testacc PKG=dynamodb TESTS=TestAccDynamoDBTag_basic
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.22.6 test ./internal/service/dynamodb/... -v -count 1 -parallel 20 -run='TestAccDynamoDBTag_basic'  -timeout 360m
=== RUN   TestAccDynamoDBTag_basic
=== PAUSE TestAccDynamoDBTag_basic
=== CONT  TestAccDynamoDBTag_basic
    tag_test.go:22: Step 1/2 error: Error running apply: exit status 1

        Error: reading dynamodb resource (arn:aws:dynamodb:us-west-2:727561393803:table/tf-acc-test-2687959590167797453) tag (key1): empty result

          with aws_dynamodb_tag.test,
          on terraform_plugin_test.tf line 28, in resource "aws_dynamodb_tag" "test":
          28: resource "aws_dynamodb_tag" "test" {

--- FAIL: TestAccDynamoDBTag_basic (19.79s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-aws/internal/service/dynamodb   26.040s
```

--- a/.changelog/39326.txt
+++ b/.changelog/39326.txt
@@ -1,0 +1,9 @@
+```release-note:bug
+resource/aws_dynamodb_tag: Handle eventual consistency of tag creation and removal
+```
+```release-note:bug
+resource/aws_dynamodb_table: Handle eventual consistency of tag creation and removal
+```
+```release-note:bug
+resource/aws_dynamodb_table_replica: Handle eventual consistency of tag creation and removal
+```

--- a/internal/service/dynamodb/generate.go
+++ b/internal/service/dynamodb/generate.go
@@ -2,7 +2,7 @@
 // SPDX-License-Identifier: MPL-2.0
 
 //go:generate go run ../../generate/tagresource/main.go
-//go:generate go run ../../generate/tags/main.go -GetTag -ListTags -ListTagsOp=ListTagsOfResource -ServiceTagsSlice -UpdateTags -ParentNotFoundErrCode=ResourceNotFoundException
+//go:generate go run ../../generate/tags/main.go -GetTag -ListTags -ListTagsOp=ListTagsOfResource -ServiceTagsSlice -UpdateTags -Wait -WaitContinuousOccurence 5 -WaitMinTimeout 1s -WaitTimeout 10m -ParentNotFoundErrCode=ResourceNotFoundException
 //go:generate go run ../../generate/servicepackage/main.go
 //go:generate go run ../../generate/listpages/main.go -ListOps=ListBackups -InputPaginator=ExclusiveStartBackupArn -OutputPaginator=LastEvaluatedBackupArn -- list_backups_pages_gen.go
 // ONLY generate directives and package declaration! Do not add anything else to this file.

--- a/internal/service/dynamodb/tags_gen.go
+++ b/internal/service/dynamodb/tags_gen.go
@@ -4,6 +4,7 @@ package dynamodb
 import (
 	"context"
 	"fmt"
+	"time"
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
@@ -165,6 +166,12 @@ func updateTags(ctx context.Context, conn *dynamodb.Client, identifier string, o
 		}
 	}
 
+	if len(removedTags) > 0 || len(updatedTags) > 0 {
+		if err := waitTagsPropagated(ctx, conn, identifier, newTags, optFns...); err != nil {
+			return fmt.Errorf("waiting for resource (%s) tag propagation: %w", identifier, err)
+		}
+	}
+
 	return nil
 }
 
@@ -172,4 +179,38 @@ func updateTags(ctx context.Context, conn *dynamodb.Client, identifier string, o
 // It is called from outside this package.
 func (p *servicePackage) UpdateTags(ctx context.Context, meta any, identifier string, oldTags, newTags any) error {
 	return updateTags(ctx, meta.(*conns.AWSClient).DynamoDBClient(ctx), identifier, oldTags, newTags)
+}
+
+// waitTagsPropagated waits for dynamodb service tags to be propagated.
+// The identifier is typically the Amazon Resource Name (ARN), although
+// it may also be a different identifier depending on the service.
+func waitTagsPropagated(ctx context.Context, conn *dynamodb.Client, id string, tags tftags.KeyValueTags, optFns ...func(*dynamodb.Options)) error {
+	tflog.Debug(ctx, "Waiting for tag propagation", map[string]any{
+		names.AttrTags: tags,
+	})
+
+	checkFunc := func() (bool, error) {
+		output, err := listTags(ctx, conn, id, optFns...)
+
+		if tfresource.NotFound(err) {
+			return false, nil
+		}
+
+		if err != nil {
+			return false, err
+		}
+
+		if inContext, ok := tftags.FromContext(ctx); ok {
+			tags = tags.IgnoreConfig(inContext.IgnoreConfig)
+			output = output.IgnoreConfig(inContext.IgnoreConfig)
+		}
+
+		return output.Equal(tags), nil
+	}
+	opts := tfresource.WaitOpts{
+		ContinuousTargetOccurence: 5,
+		MinTimeout:                1 * time.Second,
+	}
+
+	return tfresource.WaitUntil(ctx, 10*time.Minute, checkFunc, opts)
 }

